### PR TITLE
[AIRFLOW-4529] Add support for Azure Batch Service

### PR DESCRIPTION
**Pull request by ephraimbuddy**
---
This PR adds hook and operator for Azure Batch Service
https://issues.apache.org/jira/browse/AIRFLOW-4529

Issue link: [AIRFLOW-4529](https://issues.apache.org/jira/browse/AIRFLOW-4529)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/providers/microsoft/azure/hooks/azure_batch.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+                resize_errors = "\n".join([repr(e) for e in pool.resize_errors])
+                raise RuntimeError(
+                    'resize error encountered for pool {}:\n{}'.format(
+                        pool.id, resize_errors))
+            nodes = list(self.connection.compute_node.list(pool.id))
+            if (len(nodes) >= pool.target_dedicated_nodes and
+                    all(node.state in node_state for node in nodes)):
+                return nodes
+            i += 1
+            if i % 3 == 0:
```
Comment (potiuk) [inline airflow/providers/microsoft/azure/hooks/azure_batch.py]: Why the magic constant :)?

File: setup.py [hunk 2]
```diff
@@ -186,6 +186,7 @@ def write_version(filename: str = os.path.join(*[my_dir, "airflow", "git_version
     'watchtower~=0.7.3',
 ]
 azure = [
+    'azure-batch==8.0.0',
```
Comment (potiuk) [inline setup.py]: we should have >= here. By definition we should have an open-upper requirements in setup.py

**Review by potiuk — COMMENTED** [review]
Just two small comments. Great job @ephraimbuddy !

File: airflow/providers/microsoft/azure/hooks/azure_batch.py (hunk 1 above)
Comment (ephraimbuddy) [reply airflow/providers/microsoft/azure/hooks/azure_batch.py]: Mistake. I wanted to raise exception after checking the states 3 times. I can see I didn't raise it. I'll do the correction. Thanks for the review

File: setup.py (hunk 2 above)
Comment (ephraimbuddy) [reply setup.py]: Alright. Noted. I'll change it. Thanks

**Comment by potiuk** [comment]
Hey @ephraimbuddy -> please rebase now and run those `./breeze generate-requirements --python 3.6` `./breeze generate-requirements --python 3.7`

**Comment by ephraimbuddy** [comment]
> Hey @ephraimbuddy -> please rebase now and run those `./breeze generate-requirements --python 3.6` `./breeze generate-requirements --python 3.7`
Rebased and running it. Currently pulling the images and might take sometime because my machine is slow(i3).

File: airflow/providers/microsoft/azure/hooks/azure_batch.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+
+class AzureBatchHook(BaseHook):
+    """
+    Hook for Azure Batch APIs
+    """
+
+    def __init__(self, azure_batch_conn_id='azure_batch_default'):
+        super().__init__()
+        self.conn_id = azure_batch_conn_id
+        self.connection = self.get_conn()
```
Comment (kaxil) [inline airflow/providers/microsoft/azure/hooks/azure_batch.py:38]: I would say we can remove `self.connection` from here and use the get_conn or get_connection method in the `execute` method of Operator

Comment (ephraimbuddy) [reply airflow/providers/microsoft/azure/hooks/azure_batch.py:38]: Alright. I'll fix it. Thanks :)

Comment (ephraimbuddy) [reply airflow/providers/microsoft/azure/hooks/azure_batch.py:38]: I have tried to add this but I keep getting binascii.Error: Incorrect padding when I run the test for operator

**Review by potiuk — APPROVED** [review]
LGTM @kaxil - do you want to have another look ?

**Comment by kaxil** [comment]
> LGTM @kaxil - do you want to have another look ?

Checking now

**Review by kaxil — APPROVED** [review]

**Pull request merged**